game-core/src/main/java/games/strategy/triplea/delegate/EndTurnDelegate.java

```diff
@@ -206,51 +195,73 @@ private String addUnitCreatedResources(final IDelegateBridge bridge) {
     return endTurnReport.toString();
   }
 
+  /**
+   * Find all of the resources that will be created by units on the map.
+   */
+  public static IntegerMap<Resource> findUnitCreatedResources(final PlayerID player, final GameData data) {
+    final IntegerMap<Resource> resourceTotalsMap = new IntegerMap<>();
+    final Predicate<Unit> myCreatorsMatch = Matches.unitIsOwnedBy(player).and(Matches.unitCreatesResources());
+    for (final Territory t : data.getMap().getTerritories()) {
+      final Collection<Unit> myCreators = CollectionUtils.getMatches(t.getUnits().getUnits(), myCreatorsMatch);
+      for (final Unit unit : myCreators) {
+        final IntegerMap<Resource> generatedResourcesMap = UnitAttachment.get(unit.getType()).getCreatesResourcesList();
+        resourceTotalsMap.add(generatedResourcesMap);
+      }
+    }
+    for (final Resource resource : resourceTotalsMap.keySet()) {
+      if (resource.getName().equals(Constants.PUS)) {
+        resourceTotalsMap.put(resource, resourceTotalsMap.getInt(resource) * Properties.getPuMultiplier(data));
+      }
+    }
```

Comment:
I'm not sure if this for-each loop is safe.  You are iterating over `resourceTotalsMap` and then potentially modify it during the iteration.  The Javadocs for `Map#keySet()` state:

> If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation), the results of the iteration are undefined.

It might be safer to extract a separate method to do the `Resource` name lookup in the `IntegerMap` separate from applying the PU multiplier:

```java
final Resource resource = getResourceWithName(resourceTotalsMap, Constants.PUS);
if (resource != null) {
  resourceTotalsMap.put(resource, resourceTotalsMap.getInt(resource) * Properties.getPuMultiplier(data));
}

// ... somewhere ...
public static @Nullable Resource getResourceWithName(final IntegerMap<Resource> map, final String resourceName) {
  // ...
}
```

Reply:
Good catch. Actually there really isn't a need to loop through the map. Can just create a PUs resource and check for it. Updated.

Reply:
I considered that, too.  However, I hesitated to suggest it because I'm not sure if `resourceTotalsMap` gets persisted somehow.  My cause for concern is that there could then be two distinct but "equal" `Resource` objects in the serialization stream (because your new `Resource` will become part of the map during the `put()` call).  I'm not sure if there's code someplace that depends on reference equality for `Resource`s (we seem to have a big problem with reference equality when it comes to `Territory` objects).

Reply:
It should be ok as resourceTotalsMap is only a temp variable. Essentially the actual 'ChangeResourceChange' which impacts game data only checks the Resource objects by name.